File: Orimath/ViewModels/ParametricEffectDialogViewModel.cs
using System.Windows.Input;
using Orimath.Controls;
using Orimath.Plugins;
using ApplicativeProperty;

namespace Orimath.ViewModels
{
    public class ParametricEffectDialogViewModel : NotifyPropertyChanged
    {
        private readonly IParametricEffect _effect;
        private readonly IDispatcher _dispatcher;
        private readonly IMessenger _messenger;

        public string Header { get; }

        public object Parameter { get; }

        public ParametricEffectDialogViewModel(IParametricEffect effect, IDispatcher dispatcher, WorkspaceViewModel parent)
        {
            _effect = effect;
            _dispatcher = dispatcher;
            _messenger = parent;

            Header = effect.Name;
            Parameter = parent.GetEffectParameterViewModel(_effect.GetParameter());
            ExecuteCommand = Prop.True.ToCommand(Execute);
        }

        public async void Execute(object? dummy)
        {
            await _dispatcher.OnBackgroundAsync(_effect.Execute);
            _messenger.CloseDialog();
        }

        public ICommand ExecuteCommand { get; }

        public ICommand CloseCommand => _messenger.CloseDialogCommand;
    }
}

File: Orimath/ViewModels/ParametricEffectDialogViewModel.fs
namespace Orimath.ViewModels
open Orimath
open Orimath.Controls
open Orimath.Plugins
open ApplicativeProperty

type ParametricEffectDialogViewModel
    (effect: IParametricEffect,
     dispatcher: IDispatcher,
     parent: IMessenger,
     createViewModel: obj -> obj
    ) as this =
    inherit NotifyPropertyChanged()

    member val Header = Language.GetWord(effect.Name)
    member val OkText = Language.GetWord("{Ok}OK")
    member val CancelText = Language.GetWord("{Cancel}Cancel")
    member val Parameter = createViewModel (effect.GetParameter())
    member val ExecuteCommand = Prop.ctrue |> Prop.command this.Execute
    member _.CloseCommand = parent.CloseDialogCommand

    member _.Execute(_: obj) =
        dispatcher.Background {
            effect.Execute()
            dispatcher.UI.Invoke(parent.CloseDialog)
        }
